src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -30,151 +31,91 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
     /// </summary>
     public static class Monomorphize
     {
-        private struct Request
-        {
-            public QsQualifiedName OriginalName;
-            public ImmutableConcretion TypeResolutions;
-            public QsQualifiedName ConcreteName;
-        }
-
-        private struct Response
-        {
-            public QsQualifiedName OriginalName;
-            public ImmutableConcretion TypeResolutions;
-            public QsCallable ConcreteCallable;
-        }
-
         public static QsCompilation Apply(QsCompilation compilation)
         {
             var globals = compilation.Namespaces.GlobalCallableResolutions();
+            var concretizations = new List<QsCallable>();
+            var concreteNames = new Dictionary<ConcreteCallGraphNode, QsQualifiedName>();
 
-            var intrinsicCallableSet = globals
-                .Where(kvp => kvp.Value.Specializations.Any(spec => spec.Implementation.IsIntrinsic))
-                .Select(kvp => kvp.Key)
+            var nodes = new ConcreteCallGraph(compilation).Nodes
```

Comment:
This will cause a regression in monomorphization until #662 is fixed. #662 will probably be resolved by disallowing the cases that cause the bug, but the error message will not be as nice and won't show up at design time until then.

Reply:
What regression is being caused? I'm not clear on what you are referring to.

Reply:
Based on the issue linked, if the language verification were correctly forbidding the situations listed, as is the issue in question, than there would be no change as a result of this PR. The fact that the updated monomorphization more closely follows the language specifications you linked in the issue by throwing errors if the input doesn't follow it seems like it is more correct that before, as now the situations listed do not compile, which is intended. The linked issue seems to refer to a bug in the language verification where it is not invalidating programs that, by the language specifications, are supposed to be invalid.

Reply:
The regression is that those programs compile and run normally in the current version of the compiler, but cause the compiler to throw an exception in this PR. From the user's point of view, compiler exceptions look like a bug in the compiler. It would be better if there was a proper error message.

Reply:
I agree that it should be handled better, but that is a job for the language verification and not specific to the call graph or monomorphization. The situations listed should be rejected, for example, even if monomorphization is not being run as a rewrite step, as not allowing these is part of the language definition.

Reply:
My request is only that we either hold off on merging this PR until #662 is resolved, or document that this is a known issue in the release notes or somewhere else.

Reply:
Since these are edge cases that I am not sure have a real use case, I am fine with taking this in to unblock other work before they are addressed. I tentatively assigned the issue to November.